File: backend/app/core/rate_limit.py
import time
from collections import defaultdict

from fastapi import Request

from app.core.config import settings
from app.core.exceptions import AppException


class InMemoryRateLimiter:
    """Fixed-window request counter kept in this process's memory.

    Note the scope: limits are per worker process, so N workers allow N times the
    configured rate, and a restart clears all counters. That is fine as a brake on
    casual abuse, but a shared store (Redis) is what makes limits real across a
    horizontally scaled deployment.
    """

    def __init__(self, prune_interval_seconds: int = 300):
        self._history: dict[str, list[float]] = defaultdict(list)
        self._prune_interval = prune_interval_seconds
        self._last_prune = time.time()

    def is_allowed(self, key: str, max_requests: int, window_seconds: int) -> bool:
        now = time.time()
        window_start = now - window_seconds

        # Prune older records
        records = [ts for ts in self._history[key] if ts > window_start]
        self._history[key] = records

        self._prune_stale_keys(now, window_seconds)

        if len(records) >= max_requests:
            return False

        self._history[key].append(now)
        return True

    def _prune_stale_keys(self, now: float, window_seconds: int) -> None:
        """Drops keys nobody has touched recently.

        Without this, every distinct client IP would keep an entry forever, which turns
        the limiter itself into a slow memory-exhaustion target.
        """
        if now - self._last_prune < self._prune_interval:
            return
        cutoff = now - max(window_seconds, self._prune_interval)
        stale = [key for key, hits in self._history.items() if not hits or hits[-1] < cutoff]
        for key in stale:
            del self._history[key]
        self._last_prune = now


limiter = InMemoryRateLimiter()


def get_client_ip(request: Request) -> str:
    """Resolves the caller's IP, trusting only as many proxy hops as are configured.

    X-Forwarded-For is client-supplied. Taking its left-most entry lets anyone reset
    their own rate limit by sending a new fake IP on every request, so entries are read
    from the right and only TRUSTED_PROXY_COUNT of them are believed.
    """
    trusted_hops = max(0, settings.TRUSTED_PROXY_COUNT)
    direct_ip = request.client.host if request.client else "unknown"

    if trusted_hops == 0:
        return direct_ip

    forwarded = request.headers.get("X-Forwarded-For", "")
    chain = [part.strip() for part in forwarded.split(",") if part.strip()]
    if not chain:
        return direct_ip

    # The right-most entry was added by our own proxy; step back one entry per hop.
    index = len(chain) - trusted_hops
    return chain[index] if 0 <= index < len(chain) else chain[0]


def rate_limit(max_requests: int = 30, window_seconds: int = 60):
    async def dependency(request: Request):
        key = f"{request.url.path}:{get_client_ip(request)}"
        if not limiter.is_allowed(key, max_requests=max_requests, window_seconds=window_seconds):
            raise AppException(
                message="Too many requests. Please slow down and try again.",
                code="RATE_LIMIT_EXCEEDED",
                status_code=429,
            )

    return dependency

File: frontend/src/lib/rateLimit.ts
/**
 * Minimal fixed-window rate limiter for route handlers.
 *
 * Scope note: state lives in the memory of one serverless instance, so a burst spread
 * across instances gets a higher effective limit, and a cold start resets counters.
 * It is a brake on the obvious abuse (someone looping our provider-key-backed routes),
 * not a hard guarantee - that needs a shared store such as Upstash/Redis.
 */

type Bucket = { hits: number[]; };

const buckets = new Map<string, Bucket>();
let lastPrune = Date.now();

const PRUNE_INTERVAL_MS = 5 * 60 * 1000;

function prune(now: number, windowMs: number) {
  if (now - lastPrune < PRUNE_INTERVAL_MS) return;
  const cutoff = now - Math.max(windowMs, PRUNE_INTERVAL_MS);
  for (const [key, bucket] of buckets) {
    const last = bucket.hits[bucket.hits.length - 1];
    if (last === undefined || last < cutoff) buckets.delete(key);
  }
  lastPrune = now;
}

/** Returns the caller's IP as reported by the platform proxy, or a shared fallback key. */
export function getClientIp(req: Request): string {
  // On Vercel these headers are set by the platform edge, not by the client.
  const forwarded = req.headers.get("x-forwarded-for");
  if (forwarded) return forwarded.split(",")[0].trim();
  return req.headers.get("x-real-ip") ?? "unknown";
}

export function isRateLimited(key: string, maxRequests: number, windowMs: number): boolean {
  const now = Date.now();
  prune(now, windowMs);

  const bucket = buckets.get(key) ?? { hits: [] };
  const hits = bucket.hits.filter((ts) => ts > now - windowMs);

  if (hits.length >= maxRequests) {
    buckets.set(key, { hits });
    return true;
  }

  hits.push(now);
  buckets.set(key, { hits });
  return false;
}
